pkg/storage/db.go

```diff
@@ -669,3 +788,100 @@ func (db *DB) runQuery(ctx context.Context, query string, args ...interface{}) (
+          and not_before = to_timestamp(0)
+          and not_after > $1
+          and aws_resource_id = (select id from aws_resource where arn_id = $3);
+        if not found then
+            insert into aws_private_ip_assignment
+                (not_before, private_ip, aws_resource_id)
+            values ($1, $2, (select id from aws_resource where arn_id = $3));
+        end if;
+    end
+$$;
+`
+	_, err := tx.ExecContext(ctx, assignPrivateIPQuery, when, ip, arnID)
+	return err
+}
+
+func (db *DB) releasePrivateIP(ctx context.Context, tx *sql.Tx, arnID string, ip string, when time.Time) error {
+	//we use to_timestamp(0) for release events w/o known assignment as the start of epoch - 1970-01-01 is known
+	//to not have AWS resources by definition
+	//this way we can find unbalanced events while avoiding nullable not_before
+	//which provides minimal support for out-of-order events
+	const releasePrivateIPQuery = `
+do
+$$
+    begin
+        update aws_private_ip_assignment
+        set not_after=$1
+        where private_ip = $2
+          and aws_resource_id = (select id from aws_resource where arn_id = $3);
+        if not found then
+            insert into aws_private_ip_assignment
+                (not_before, not_after, private_ip, aws_resource_id)
+            values (to_timestamp(0), $1, $2, (select id from aws_resource where arn_id = $3));
+        end if;
+    end
+$$;
+`
+	_, err := tx.ExecContext(ctx, releasePrivateIPQuery, when, ip, arnID)
+	return err
+}
+
+func (db *DB) assignPublicIP(ctx context.Context, tx *sql.Tx, arnID string, ip string, hostname string, when time.Time) error {
+	const assignPublicIPQuery = `
+do
+$$
+    begin
+        update aws_public_ip_assignment
+        set not_before = $1
+        where public_ip = $2
+          and not_before = to_timestamp(0)
+          and not_after > $1
+          and aws_resource_id = (select id from aws_resource where arn_id = $3)
+          and aws_hostname = $4;
+        if not found then
+            insert into aws_public_ip_assignment
+                (not_before, public_ip, aws_resource_id, aws_hostname)
+            values ($1, $2, (select id from aws_resource where arn_id = $3), $4);
+        end if;
+    end
+$$;
+`
+	_, err := tx.ExecContext(ctx, assignPublicIPQuery, when, ip, arnID, hostname)
+	return err
+}
+
+func (db *DB) releasePublicIP(ctx context.Context, tx *sql.Tx, arnID string, ip string, hostname string, when time.Time) error {
+	//we use to_timestamp(0) for release events w/o known assignment as the start of epoch - 1970-01-01 is known
+	//to NOT have AWS resources by definition
+	//this way we can find unbalanced events while avoiding nullable not_before
+	//which provides minimal support for out-of-order events
+	const releasePublicIPQuery = `
+do
+$$
+    begin
+        update aws_public_ip_assignment
+        set not_after=$1
+        where public_ip = $2
+          and aws_resource_id = (select id from aws_resource where arn_id = $3);
```

Comment:
does the update statement also need to match on hostname?

Reply:
The ips and hostnames always "match", as amazon generates them. I was looking into the old db which has assign/release events recorded separately.
On the other hand, I do not see a reason to NOT match for clarity/cleanliness so, will update :)

Reply:
Added